Mathlib/Combinatorics/SimpleGraph/Diameter.lean

```diff
@@ -0,0 +1,206 @@
+/-
+Copyright (c) 2024 Rida Hamadani. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Rida Hamadani
+-/
+import Mathlib.Combinatorics.SimpleGraph.Metric
+
+/-!
+# Diameter of a simple graph
+
+This module defines the diameter of a simple graph, which measures the maximum distance between
+vertices.
+
+## Main definitions
+
+- `SimpleGraph.ediam` is the graph extended diameter.
+
+- `SimpleGraph.diam` is the graph diameter.
+
+## Todo
+
+- Prove that `G.egirth ≤ 2 * G.ediam + 1` and `G.girth ≤ 2 * G.diam + 1` when the diameter is
+  non-zero.
+
+-/
+
+namespace SimpleGraph
+variable {α : Type*} {G G' : SimpleGraph α}
+
+section ediam
+
+/--
+The extended diameter is the greatest distance between any two vertices, with the value `⊤` in
+case the distances are not bounded above, or the graph is not connected.
+-/
+noncomputable def ediam (G : SimpleGraph α) : ℕ∞ :=
+  ⨆ u, ⨆ v, G.edist u v
+
+lemma ediam_def : G.ediam = ⨆ p : (α × α), G.edist p.1 p.2 := by
+  rw [ediam, iSup_prod]
+
+lemma edist_le_ediam {u v : α} : G.edist u v ≤ G.ediam :=
+  LE.le.trans (le_iSup _ _) <| le_iSup_iff.mpr fun _ h ↦ h u
+
+lemma ediam_eq_top_iff : G.ediam = ⊤ ↔ ∀ b < ⊤, ∃ u v, b < G.edist u v := by
+  simp only [ediam, iSup_eq_top, lt_iSup_iff]
+
+lemma ediam_eq_zero_of_isEmpty [IsEmpty α] : G.ediam = 0 := by
+  rw [ediam, ciSup_of_empty, bot_eq_zero']
+
+lemma nonempty_of_ediam_ne_zero (h : G.ediam ≠ 0) : Nonempty α := by
```

Comment:
```suggestion
lemma nontrivial_of_ediam_ne_zero (h : G.ediam ≠ 0) : Nontrivial α := by
```

Reply:
The point of having this is that there are many lemmas in this API which require `Nonempty α` as a typeclass assumption, so I thought this would provide an easy way to have that assumption.
What do you think of turning `zero_lt_ediam_of_nontrivial` into an iff statement instead?

Reply:
Having it as an iff statement is good, but we should also have the alias for the nontrivial => ... and subsingleton => ... versions, where the assumption is a typeclass assumption

Reply:
Well, typeclass search knows that `Nontrivial α` implies `Nonempty α`, so I don't really see the problem?

Reply:
Ah, good to know! Working on it.

Reply:
For clarity, I was replying to Rida's message above